### ISSUE TITLE
NBS: Some edge case leads snappy to allocate output slice

### ISSUE THREAD

**Issue by cmasone-attic** [attic-labs/noms#3156]
This is the source of the "corruption" in #3148 

For some large chunks, `snappy.Encode` decides to allocate some new output storage instead of using the buffer we pass it. We were assuming that snappy DID use the buffer we passed it, so this resulted in big blocks of 0's in the table file and chunks being left out (!!!). One such table is `767sh0tj3daui5rh6tj9nb21iu1it2df`, which github won't let me attach :-/

Even using what's in that broken table, I can't craft a test case that reproduces the problem. Unfortunately, the problematic chunks aren't in there.

I'm going to modify the code to detect this case and manually copy the compressed chunk data into the table, but also ship the uncompressed data off to storage somewhere for analysis.

**Comment by cmasone-attic** [comment]
We don't understand why yet, but we're still seeing some of this. Working around for now

**Comment by cmasone-attic** [comment]
We figured it out!

When trying to calculate the maximum possible table size for a compacted table, we were using the _compressed_ lengths of chunks in the tables being compacted -- because that's the only information we had! We realized, though, that we could repurpose a currently-unused field in the NBS table footer format. So, now NBS will store the total _uncompressed_ length of all stored chunks (as opposed to the total _compressed_ length, which it used to store) in the table footer. This allows compaction code to correctly estimate how much space it'll take to store all those chunks in a freshly compacted table with a bunch of other chunk data.
